jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/client/GumbyClient.kt

```diff
@@ -37,64 +49,124 @@ class GumbyClient(private val project: Project) {
+                    .uploadId(uploadId)
+            }
+            .transformationSpec { spec ->
+                spec.transformationType(TransformationType.LANGUAGE_UPGRADE)
+                    .source { it.language(sourceLanguage) }
+                    .target { it.language(targetLanguage) }
+            }
+            .build()
+        return callApi({ bearerClient().startTransformation(request) }, apiName = CodeTransformApiNames.StartTransformation, uploadId = uploadId)
     }
 
-    fun getCodeModernizationPlan(jobId: JobId): GetTransformationPlanResponse = bearerClient().getTransformationPlan {
-        it.transformationJobId(jobId.id)
+    fun getCodeModernizationPlan(jobId: JobId): GetTransformationPlanResponse {
+        val request = GetTransformationPlanRequest.builder().transformationJobId(jobId.id).build()
+        return callApi({ bearerClient().getTransformationPlan(request) }, apiName = CodeTransformApiNames.GetTransformationPlan, jobId = jobId.id)
     }
 
-    fun stopTransformation(transformationId: String): StopTransformationResponse = bearerClient().stopTransformation {
-        it.transformationJobId(transformationId)
+    fun stopTransformation(transformationJobId: String): StopTransformationResponse {
+        val request = StopTransformationRequest.builder().transformationJobId(transformationJobId).build()
+        return callApi({ bearerClient().stopTransformation(request) }, apiName = CodeTransformApiNames.StopTransformation, jobId = transformationJobId)
+    }
+
+    private fun <T : CodeWhispererRuntimeResponse> callApi(
+        apiCall: () -> T,
+        apiName: CodeTransformApiNames,
+        jobId: String? = null,
+        uploadId: String? = null,
+    ): T {
+        val startTime = Instant.now()
+        var result: CodeWhispererRuntimeResponse? = null
+        try {
+            result = apiCall()
+            return result
+        } catch (e: Exception) {
+            LOG.error(e) { "$apiName failed: ${e.message}" }
+            CodetransformTelemetry.logApiError(
+                codeTransformApiErrorMessage = e.message.toString(),
+                codeTransformApiNames = apiName,
+                codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
+                codeTransformJobId = jobId,
+            )
+            throw e // pass along error to callee
+        } finally {
+            CodetransformTelemetry.logApiLatency(
+                codeTransformApiNames = apiName,
+                codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
+                codeTransformRunTimeLatency = calculateTotalLatency(startTime, Instant.now()),
+                codeTransformUploadId = uploadId,
+                codeTransformJobId = jobId,
+                codeTransformRequestId = result?.responseMetadata()?.requestId(),
+            )
+        }
     }
 
     suspend fun downloadExportResultArchive(jobId: JobId): MutableList<ByteArray> {
```

Comment:
Can we not call this API via the generic function just above?

Reply:
downloadExportResultArchive doesn't return a CodeWhispererRuntimeResponse, and has a lot more logic going on inside it, so I decided to keep it separate.